### PR TITLE
[HPO] Optional ray lightning and backup implementation

### PR DESCRIPTION
**Pull request by yinweisu**
*Issue #, if available:*
https://github.com/awslabs/autogluon/issues/1798

*Description of changes:*
* Change ray lightning to be optional because the support level of it remains unclear. Current strategy is to let user install ray lightning and the corresponding pytorch lightning themselves. We can consider adding ray lightning back in our `setup.py` after they support pytorch lightning 1.6, which is mandatory for 0.5 release(https://github.com/awslabs/autogluon/issues/1831).
* Without ray lightning, we can still run trials in parallel but each individual trial won't be able to utilize multiple gpus.
* Updated ray version for protobuf and hyperopt fix.
* Extracted common logic of each adapter to abstract adapter.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1828-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1828/3/index.html

**Comment by szha** [comment]
Job PR-1828-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1828/4/index.html

File: core/setup.py [hunk 1]
```diff
@@ -38,22 +38,18 @@
...
-        'ray_lightning>=0.2.0,<0.3.0',
+        'ray[tune]>=1.13,<1.14',
         'hyperopt>=0.2.7,<0.2.8',
         # 'GPy>=1.10.0,<1.11.0'  # TODO: Enable this once PBT/PB2 are supported by ray lightning
     ]
 }
 
 tests_require = [
     'pytest',
+    'ray_lightning',  # test ray lightning resource calculation
```
Comment (Innixma) [inline core/setup.py:52]: Missing version?

**Review by Innixma — APPROVED** [review]
LGTM! Had one comment

**Comment by szha** [comment]
Job PR-1828-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1828/5/index.html

**Review by zhiqiangdon — APPROVED** [review]

**Review by sxjscience — APPROVED** [review]

**Pull request merged**